standalone/lcio2edm4hep.cpp

```diff
@@ -37,52 +39,124 @@ std::vector<std::pair<std::string, std::string>> getNamesAndTypes(const std::str
+  -n N              Limit the number of events to convert to N (default = -1, all events)
+
+Examples:
+- print this message:
+lcio2edm4hep -h
+- convert complete file (needs all lcio collections to be present in all files):
+lcio2edm4hep infile.slcio outfile_edm4hep.root
+- the same but providing complete set of collections (either to patch collections in,
+  or to only convert a subset):
+lcio2edm4hep infile.slcio outfile_edm4hep.root coltype.txt
 )";
 
-int main(int argc, char* argv[])
+struct ParsedArgs {
+  std::string inputFile {};
+  std::string outputFile {};
+  std::string patchFile {};
+  int nEvents {-1};
+};
+
+void printUsageAndExit()
+{
+  std::cerr << usageMsg << std::endl;
+  std::exit(1);
+}
+
+ParsedArgs parseArgs(std::vector<std::string> argv)
 {
-  if (argc == 2 && (argv[1] == std::string("-h") || argv[1] == std::string("--help"))) {
+  // find help
+  if (std::find_if(argv.begin(), argv.end(), [](const auto& elem) {
+        return elem == "-h" || elem == "--help";
+      }) != argv.end()) {
     std::cerr << usageMsg << '\n' << helpMsg << std::endl;
-    return 0;
+    std::exit(0);
   }
+
+  auto argc = argv.size();
+  if (argc < 3 || argc > 7) {
+    printUsageAndExit();
+  }
+
+  ParsedArgs args;
+  auto nEventIt = std::find(argv.begin(), argv.end(), "-n");
+  if (nEventIt != argv.end()) {
+    const auto index = std::distance(argv.begin(), nEventIt);
+    if (index > argc - 2) {
+      // No argument left to parse
+      printUsageAndExit();
+    }
+    const auto& value = argv[index + 1]; // get the actual value
+    try {
+      args.nEvents = std::stoi(value);
+    } catch (std::invalid_argument& err) {
+      std::cerr << "Cannot parse " << value << " as an integer" << std::endl;
+      printUsageAndExit();
+    }
+    argv.erase(nEventIt, nEventIt + 2);
+  }
+
+  argc = argv.size();
   if (argc < 3) {
-    std::cerr << usageMsg << std::endl;
-    return 1;
+    printUsageAndExit();
   }
-  const auto outputFile = std::string(argv[2]);
+  args.inputFile = argv[1];
+  args.outputFile = argv[2];
+  if (argc == 4) {
+    args.patchFile = argv[3];
+  }
+  return args;
+}
+
+int main(int argc, char* argv[])
+{
+  const auto args = parseArgs({argv, argv + argc});
 
-  bool patching = false;
   UTIL::CheckCollections colPatcher {};
-  if (argc == 4) {
-    const auto namesTypes = getNamesAndTypes(argv[3]);
+  std::vector<std::pair<std::string, std::string>> namesTypes {};
+  const bool patching = !args.patchFile.empty();
+  if (patching) {
+    namesTypes = getNamesAndTypes(args.patchFile);
     if (namesTypes.empty()) {
       std::cerr << "The provided list of collection names and types does not satisfy the required format: Pair of Name "
                    "and Type per line separated by space"
                 << std::endl;
       return 1;
     }
     colPatcher.addPatchCollections(namesTypes);
-    patching = true;
   }
+  // Construct a vector of collections to convert. If namesTypes is empty, this
```

Comment:
That is only necessary for the patching of the `LCEvent`. We pass it to the `colPatcher` a few lines above and that stores it internally and uses it in `patchCollections`. For `convertEvent` we only need to know the names of the collections we want to convert, because we have to obtain the type in any case (to cover the cases where the user doesn't provide us with this optional input).